executor/slow_query.go

```diff
@@ -465,3 +519,161 @@ func ParseTime(s string) (time.Time, error) {
+		}
+		//All rotated log files have the same prefix with the original file
+		if !strings.HasPrefix(path, prefix) {
+			return nil
+		}
+		file, err := os.OpenFile(path, os.O_RDONLY, os.ModePerm)
+		if err != nil {
+			return handleErr(err)
+		}
+		skip := false
+		defer func() {
+			if !skip {
+				terror.Log(file.Close())
+			}
+		}()
+		// Get the file start time.
+		fileBeginTime, err := e.getFileStartTime(file)
+		if err != nil {
+			return handleErr(err)
+		}
+		if fileBeginTime.After(e.extractor.EndTime) {
+			return nil
+		}
+
+		// Get the file end time.
+		fileEndTime, err := e.getFileEndTime(file)
+		if err != nil {
+			return handleErr(err)
+		}
+		if fileEndTime.Before(e.extractor.StartTime) {
+			return nil
+		}
+		_, err = file.Seek(0, io.SeekStart)
+		if err != nil {
+			return handleErr(err)
+		}
+		logFiles = append(logFiles, logFile{
+			file:  file,
+			begin: fileBeginTime,
+			end:   fileEndTime,
+		})
+		skip = true
+		return nil
+	})
+	// Sort by start time
+	sort.Slice(logFiles, func(i, j int) bool {
+		return logFiles[i].begin.Before(logFiles[j].begin)
+	})
+	return logFiles, err
+}
+
+func (e *SlowQueryRetriever) getFileStartTime(file *os.File) (time.Time, error) {
+	var t time.Time
+	_, err := file.Seek(0, io.SeekStart)
+	if err != nil {
+		return t, err
+	}
+	reader := bufio.NewReader(file)
+	maxNum := 128
+	for {
+		lineByte, err := getOneLine(reader)
+		if err != nil {
+			return t, err
+		}
+		line := string(lineByte)
+		if strings.HasPrefix(line, variable.SlowLogStartPrefixStr) {
+			return ParseTime(line[len(variable.SlowLogStartPrefixStr):])
+		}
+		maxNum -= 1
+		if maxNum <= 0 {
+			break
+		}
+	}
+	return t, errors.Errorf("can't found start time")
+}
+func (e *SlowQueryRetriever) getFileEndTime(file *os.File) (time.Time, error) {
+	var t time.Time
+	stat, err := file.Stat()
+	if err != nil {
+		return t, err
+	}
+	fileSize := stat.Size()
+	cursor := int64(0)
+	line := make([]byte, 0, 64)
+	maxNum := 128
+	for {
+		cursor -= 1
+		_, err := file.Seek(cursor, io.SeekEnd)
+		if err != nil {
+			return t, err
+		}
+
+		char := make([]byte, 1)
+		_, err = file.Read(char)
+		if err != nil {
+			return t, err
+		}
+		// If find a line.
+		if cursor != -1 && (char[0] == 10 || char[0] == 13) {
+			for i, j := 0, len(line)-1; i < j; i, j = i+1, j-1 {
+				line[i], line[j] = line[j], line[i]
+			}
+			lineStr := string(line)
+			lineStr = strings.TrimSpace(lineStr)
+			if strings.HasPrefix(lineStr, variable.SlowLogStartPrefixStr) {
+				return ParseTime(lineStr[len(variable.SlowLogStartPrefixStr):])
+			}
+			line = line[:0]
+			maxNum -= 1
+		}
+		line = append(line, char[0])
+		if cursor == -fileSize || maxNum <= 0 {
+			return t, errors.Errorf("can't found end time")
+		}
+	}
+}
+
+// NewSlowQueryRetrieverForTest was only used in test.
+func NewSlowQueryRetrieverForTest(extractor *plannercore.SlowQueryExtractor, files []logFile) *SlowQueryRetriever {
+	return &SlowQueryRetriever{extractor: extractor, files: files}
```

Comment:
I think we should export the `SlowQueryRetriever` and it's better to move the `NewSlowQueryRetrieverForTest` to file `executor_pkg_test.go`.

Reply:
great, done.